payu/metadata.py

```diff
@@ -0,0 +1,269 @@
+"""Payu experiment UUID and metadata support
+
+Generates and commit a new experiment uuid and updates/creates experiment
+metadata
+
+:copyright: Copyright 2011 Marshall Ward, see AUTHORS for details.
+:license: Apache License, Version 2.0, see LICENSE for details.
+"""
+
+import warnings
+from pathlib import Path
+from typing import Optional, List
+
+import shortuuid
+from ruamel.yaml import YAML
+from ruamel.yaml.comments import CommentedMap
+
+from payu.fsops import read_config
+from payu.laboratory import Laboratory
+from payu.git_utils import get_git_branch, get_git_user_info, git_commit
+
+# A truncated uuid is used for branch-uuid aware experiment names 
+TRUNCATED_UUID_LENGTH = 5
+METADATA_FILENAME = 'metadata.yaml'
+
+USAGE_HELP = """
+If this is a new experiment, create a new git branch by running:
+    payu checkout -b NEW_BRANCH_NAME
+where NEW_BRANCH_NAME is name of the new branch. This will generate a new
+uuid, a branch-uuid aware experiment name and commit changes to the
+metadata file.
+
+Alternatively to generate a new uuid or experiment name on the current git 
+branch at the next payu setup or run command, remove the pre-existing uuid or
+experiment name from the metadata file.
+
+Note: Experiment names are the name used for work and archive directories 
+in the laboratory directory.
+"""
+
+
+class ExperimentMetadataError(Exception):
+    """Class for experiment name exceptions"""
+    def __init__(self, message="Invalid experiment name in metadata"):
+        super().__init__(message)
+        print(USAGE_HELP)
+
+class MetadataWarning(Warning):
+    pass
+
+class Metadata:
+    """
+    Class to store/update/create metadata such as experiment uuid and name
+
+    Parameters:
+        lab : Laboratory
+            The modules laboratory
+        branch : str | None = None
+            The git branch on which the experiment is run
+        control_path : Path | None = None
+            Path to where the experiment is configured and run. The default
+            is set to the current working directory. This default is set in
+            in fsops.read_config
+        config_path : Path | None = None
+            Configuration Path. The default is config.yaml in the current
+            working directory. This is also set in fsop.read_config
+    """
+
+    def __init__(self,
+                 lab: Laboratory,
+                 config_path: Optional[Path] = None,
+                 branch: Optional[str] = None,
+                 control_path: Optional[Path] = None) -> None:
+        self.lab = lab
+        self.config = read_config(config_path)
+
+        if control_path is None:
+            control_path = Path(self.config.get('control_path'))
```

Comment:
`get_control_path` in `branch.py` reads config file and then just returns control path, while metadata class needs to use the saved config in a couple places. To avoid the control directory being defined in metadata, I could pass the control directory directory directly to metadata class in both the experiment and in `branch.py`. Or have `metadata.py` be the only file where control path is read from config - so branch and experiment use the same one.. Somewhat related question - do you know of any cases where control path is set in config.yaml and it's not the same directory that config.yaml resides in?

Reply:
> To avoid the control directory being defined in metadata, I could pass the control directory directory directly to metadata class in both the experiment and in `branch.py`. Or have `metadata.py` be the only file where control path is read from config - so branch and experiment use the same one

Oh, I seem to have opened a can of worms. I don't think you should change it because of what I said. I just noticed you'd make a function elsewhere, but you've explained the reasoning which seems fine.

> do you know of any cases where control path is set in config.yaml and it's not the same directory that config.yaml resides in?

No. It isn't even really obvious to me what the use case might be for this.

Reply:
Ok I'll leave that for that for another day perhaps